File: php/src/Relation.php
<?php

declare(strict_types=1);

namespace LiteDbModel\Runtime;

/**
 * litedbmodel v2 SCP — read-relation batch EXECUTION (PHP port of src/scp/relation.ts, #43).
 *
 * Byte-for-byte port of the TS reference relation runtime: the STATIC pre-compiled batch op
 * (`bundle->relations->{name}` — pure JSON) is EXECUTED, never regenerated. A RelationOp carries
 * the batched child SELECT text with ONE `?` for the deduped-key array param; the runtime dedupes
 * the parent keys, resolves the deferred PG array cast from the REAL keys, renders `?`→`$N`, short-
 * circuits an empty key set (NO query), runs the batch, groups the child rows by target key, and
 * distributes them onto the parents per cardinality (hasMany → list, belongsTo/hasOne → single or
 * null). The SAME runRelationOp / distributeToParent / dedupeKeys the TS eager path uses.
 *
 * The PG array binds as the `{…}` array-literal text at the read seam (StaticBundle::pgArrayLiteral),
 * matching the existing #46 IN-list binding — the `::int[]` cast is resolved from the real keys, so
 * PG receives `= ANY($1::int[])`. MySQL/SQLite bind the JSON-encoded array string. Grouped-then-
 * distributed by key → deterministic regardless of #40 sibling-relation completion order.
 */
final class Relation
{
    private const PG_ARRAY_CAST_TOKEN = '@@PG_ARRAY_CAST@@';

    /**
     * The ordered PARENT / CHILD key columns (single-key → 1-element; composite → the tuple, #47 item 1).
     *
     * @return list<string>
     */
    private static function parentKeyCols(\stdClass $op): array
    {
        if (isset($op->parentKeys) && is_array($op->parentKeys)) {
            return array_map('strval', $op->parentKeys);
        }
        return [(string) $op->parentKey];
    }

    /** @return list<string> */
    private static function targetKeyCols(\stdClass $op): array
    {
        if (isset($op->targetKeys) && is_array($op->targetKeys)) {
            return array_map('strval', $op->targetKeys);
        }
        return [(string) $op->targetKey];
    }

    /**
     * The deduped, non-null parent-key TUPLES (insertion order preserved). Thin delegator to the
     * shared grouping core ({@link Grouping::dedupeKeyTuples}) — the SSoT for the drop/dedupe/tuple
     * semantics (no duplicated grouping logic).
     *
     * @param list<\stdClass> $parents
     * @param list<string> $keyCols
     * @return list<list<mixed>>
     */
    public static function dedupeKeys(array $parents, array $keyCols): array
    {
        return Grouping::dedupeKeyTuples($parents, $keyCols);
    }

    /**
     * Bind the deduped keys to the op's params per dialect + arity (mirrors TS bindKeys). Single-key:
     * PG → ONE `{…}` array-literal param; MySQL/SQLite → ONE JSON scalar-array string. Composite: PG →
     * ONE `{…}` array-literal PER key column (transposed tuples); MySQL/SQLite → ONE JSON
     * array-of-tuples string. Returns the positional param list.
     *
     * @param list<list<mixed>> $tuples
     * @return list<string>
     */
    private static function bindKeys(\stdClass $op, array $tuples): array
    {
        $composite = isset($op->parentKeys);
        if ((string) $op->dialect === 'postgres') {
            $nCols = $composite ? count(self::parentKeyCols($op)) : 1;
            $args = [];
            for ($col = 0; $col < $nCols; $col++) {
                $colArr = array_map(static fn ($t) => $t[$col], $tuples);
                $args[] = StaticBundle::pgArrayLiteral($colArr);
            }
            return $args;
        }
        $payload = $composite
            ? array_map(static fn ($t) => array_values($t), $tuples)
            : array_map(static fn ($t) => $t[0], $tuples);
        return [json_encode($payload, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE)];
    }

    /**
     * Run ONE relation batch op for a set of parent rows (byte-for-byte port of TS runRelationOp).
     * Dedup the parent-key tuples, resolve the deferred PG array cast(s) from the REAL keys (one per
     * key column for composite) BEFORE the `?`→`$N` render; on a NON-empty key set execute binding
     * the keys (single array / per-column arrays / JSON tuples) and group the child rows by
     * target-key identity. An EMPTY key set issues NO query.
     *
     * @param list<\stdClass> $parents
     * @return array<string, list<\stdClass>>
     */
    public static function runRelationOp(\stdClass $op, array $parents, \PDO|ExecutionContext $db): array
    {
        $ctx = Context::of($db);
        $dialect = (string) $op->dialect;
        $pCols = self::parentKeyCols($op);
        $keys = self::dedupeKeys($parents, $pCols);
        $batch = [];
        $sql = (string) $op->sql;
        if ($dialect === 'postgres') {
            foreach ($pCols as $col => $_) {
                $colVals = array_map(static fn ($t) => $t[$col], $keys);
                $sql = StaticBundle::resolvePgArrayCast($sql, $colVals);
            }
        }
        $sql = StaticBundle::renderPlaceholders($sql, $dialect);
        if (count($keys) === 0) {
            return $batch;
        }
        $tCols = self::targetKeyCols($op);
        // The central READ seam (§2): the ONE driver contact for the relation batch. Byte-identical to
        // the pre-seam `$db->prepare($sql)->execute(bindKeys)->fetchAll(OBJ)`.
        $rows = execute($ctx, $sql, self::bindKeys($op, $keys));
        // Hard-limit runaway guard (Phase E-2, epic #74; v1 `_selectForRelation`): POST-fetch, if the
        // batch TOTAL exceeds the baked cap, throw with the EXACT count (the batch is fetched in full,
        // no N+1). ⚠️ field mapping mirrors the TS reference: `model` = the relation TARGET TABLE,
        // `relation` = the relation NAME. ABSENT `op->hardLimit` ⇒ NO check (disabled / an intrinsic
        // per-parent-`limit` relation whose fanout is already bounded). Thrown BEFORE grouping so an
        // over-cap read never assembles an unbounded result set. Cap read from the artifact ONLY.
        $rowList = is_array($rows) ? $rows : [];
        if (isset($op->hardLimit)) {
            // The relation-context arm of the shared runaway check (SSoT) — the SAME `count > limit ⇒
            // throw` primitive, so the comparison + error assembly live in ONE place (mirror of
            // python relation.py). ABSENT `op->hardLimit` ⇒ NO check.
            LimitExceededError::check(
                (int) $op->hardLimit,
                count($rowList),
                'relation',
                isset($op->targetTable) ? (string) $op->targetTable : null,
                (string) $op->name,
            );
        }
        // Group the child rows by target-key identity via the shared core (a null/absent target key
        // drops the child — the SSoT drop semantics).
        return Grouping::groupByKey($rowList, $tCols);
    }

    /**
     * Distribute a resolved batch onto ONE parent per cardinality (port of TS distributeToParent):
     * hasMany → the child list ([] when none); belongsTo/hasOne → the single child (or null). Keyed
     * by the parent's key-tuple identity.
     *
     * @param array<string, list<\stdClass>> $batch
     * @return list<\stdClass>|\stdClass|null
     */
    public static function distributeToParent(\stdClass $op, \stdClass $parent, array $batch): mixed
    {
        // Delegate to the shared grouping core; cardinality maps onto its `single` flag (hasMany →
        // list, belongsTo/hasOne → single or null). No duplicated grouping logic.
        $single = (string) $op->kind !== 'hasMany';
        return Grouping::attachToParent($parent, self::parentKeyCols($op), $batch, $single);
    }

    /**
     * The driver a relation runs against: its tagged cross-DB connection, else the primary $db.
     * CROSS-DB (V0 R1): a relation whose op carries a `connection` tag (its target model lives in a
     * DIFFERENT DB — v1 LazyRelation.ts:236) routes to $connections[tag]. Loud failure when the tag
     * has no registered driver (a real wiring bug — never a silent same-DB fallback that would run
     * the target's query on the wrong DB). Untagged relations use the primary $db.
     *
     * @param array<string,ExecutionContext> $connections
     */
    private static function driverForOp(\stdClass $op, ExecutionContext $ctx, array $connections): ExecutionContext
    {
        $tag = $op->connection ?? null;
        if ($tag === null) {
            return $ctx;
        }
        if (!isset($connections[$tag])) {
            throw new \RuntimeException(
                "cross-DB relation '" . ($op->name ?? '?') . "': no driver registered for connection "
                . "'{$tag}' (pass it in readBundle connections)"
            );
        }
        return $connections[$tag];
    }

    /**
     * Run a READ bundle's primary row list, then batch-load + hydrate the selected relations onto
     * each parent (port of the TS readBundle typed-object surface, declarative-select path). The
     * primary read output must be a bare row list; each named relation in $withNames is batch-
     * prefetched ONCE over the whole page (staged, no N+1) via the SAME runRelationOp and attached
     * onto each parent as an own property.
     *
     * CROSS-DB (V0 R1): a relation op carrying a `connection` tag is batched against
     * $connections[tag] (its target model's DB) instead of the primary $db; untagged relations
     * ignore $connections. Pass an empty array for a single-DB read.
     *
     * @param array<string,mixed> $input
     * @param list<string> $withNames
     * @param array<string,\PDO|ExecutionContext> $connections
     * @return list<\stdClass>
     */
    public static function readBundle(\stdClass $bundle, array $input, \PDO|ExecutionContext $db, array $withNames, array $connections = []): array
    {
        $ctx = Context::of($db);
        $connCtx = array_map(static fn (\PDO|ExecutionContext $c): ExecutionContext => Context::of($c), $connections);
        $out = Runtime::executeBundle($bundle, $input, $ctx);
        if (!is_array($out) || !array_is_list($out)) {
            throw new \RuntimeException(
                'scp read: the read behavior output is not a row list; the typed-object read surface '
                . 'expects a Select-shaped output'
            );
        }
        /** @var list<\stdClass> $rows */
        $rows = $out;
        $relations = $bundle->relations ?? null;
        foreach ($withNames as $name) {
            if (!($relations instanceof \stdClass) || !property_exists($relations, $name)) {
                throw new \RuntimeException("declarative select: relation '{$name}' is not declared on this model");
            }
            self::hydrateRelation($relations->{$name}, $rows, $ctx, $connCtx, $name);
        }
        return $rows;
    }

    /**
     * Hydrate ONE relation edge over $parents (ONE batched query, N+1-free), then RECURSE into
     * $op->childRelations — the batched-map-over-batched-map chain the native codegen path lowers,
     * reproduced for the runtime/ir-exec path.
     *
     * One edge = one query, INDEPENDENT of the parent count: runRelationOp dedupes the parent keys and
     * fetches ALL children with ONE `WHERE fk IN (…)` batch, then the grouping SSoT nests them onto each
     * parent via distributeToParent. A nested level batches over the FLATTENED child rows fetched here —
     * the EXACT objects attached to the parents (PHP object handles), so grandchildren hydrate in place
     * (users→posts→comments = 3 queries, not 1 + N + N·M). No new mechanism: every level runs the SAME
     * runRelationOp + grouping core.
     *
     * @param list<\stdClass>              $parents
     * @param array<string,ExecutionContext> $connCtx
     */
    private static function hydrateRelation(\stdClass $op, array $parents, ExecutionContext $ctx, array $connCtx, string $attachName): void
    {
        $relCtx = self::driverForOp($op, $ctx, $connCtx);
        $batch = self::runRelationOp($op, $parents, $relCtx);
        foreach ($parents as $p) {
            $p->{$attachName} = self::distributeToParent($op, $p, $batch);
        }
        $childOps = $op->childRelations ?? null;
        if (is_array($childOps) && $childOps !== []) {
            // The flattened child rows (each child appears ONCE, keyed by its target tuple) = the next
            // level's parent set. Empty ⇒ no grandchild query (short-circuit, still N+1-free).
            $childRows = [];
            foreach ($batch as $group) {
                foreach ($group as $child) {
                    $childRows[] = $child;
                }
            }
            if ($childRows !== []) {
                foreach ($childOps as $childOp) {
                    self::hydrateRelation($childOp, $childRows, $ctx, $connCtx, (string) $childOp->name);
                }
            }
        }
    }
}

File: rust/litedbmodel_interpreter/src/relation.rs
//! Interpreter-only adapter from published relation metadata to the native relation batch core.

use std::collections::HashMap;

use behavior_contracts::Value;

use crate::codegen_exec::ArrayParamShape;
use crate::driver::Driver;
use crate::errors::{RuntimeError, SqlFailure};
use crate::node::Node;
use crate::runtime::execute_bundle;
use litedbmodel_runtime::relation::{
    execute_relation_batch, group_children, matched_children, RelationBatch,
};

struct RelationOp {
    name: String,
    kind: String,
    parent_keys: Vec<String>,
    target_keys: Vec<String>,
    key_shape: ArrayParamShape,
    connection: Option<String>,
    sql: String,
    target_table: Option<String>,
    hard_limit: Option<i64>,
}

fn string(node: &Node, key: &str) -> String {
    node.get(key)
        .and_then(Node::as_str)
        .unwrap_or("")
        .to_string()
}

fn keys(node: &Node, many: &str, one: &str) -> Vec<String> {
    node.get(many)
        .and_then(Node::as_array)
        .map(|values| {
            values
                .iter()
                .filter_map(Node::as_str)
                .map(str::to_string)
                .collect()
        })
        .unwrap_or_else(|| vec![string(node, one)])
}

fn op_from_node(node: &Node) -> RelationOp {
    RelationOp {
        name: string(node, "name"),
        kind: string(node, "kind"),
        parent_keys: keys(node, "parentKeys", "parentKey"),
        target_keys: keys(node, "targetKeys", "targetKey"),
        key_shape: match node.get("keyShape").and_then(Node::as_str) {
            Some("per_column") => ArrayParamShape::PerColumn,
            _ => ArrayParamShape::SingleJson,
        },
        connection: node
            .get("connection")
            .and_then(Node::as_str)
            .map(str::to_string),
        sql: string(node, "sql"),
        target_table: node
            .get("targetTable")
            .and_then(Node::as_str)
            .map(str::to_string),
        hard_limit: node.get("hardLimit").and_then(Node::as_i64),
    }
}

fn field<'a>(row: &'a Value, key: &str) -> Option<&'a Value> {
    match row {
        Value::Obj(fields) => fields
            .iter()
            .find(|(name, _)| name == key)
            .map(|(_, value)| value),
        _ => None,
    }
}

fn tuples(rows: &[Value], keys: &[String]) -> Vec<Vec<Value>> {
    rows.iter()
        .map(|row| {
            keys.iter()
                .map(|key| field(row, key).cloned().unwrap_or(Value::Null))
                .collect()
        })
        .collect()
}

fn fetch(
    op: &RelationOp,
    parents: &[Vec<Value>],
    driver: &dyn Driver,
) -> Result<Vec<Value>, RuntimeError> {
    let columns = (0..op.parent_keys.len())
        .map(|column| parents.iter().map(|row| row[column].clone()).collect())
        .collect::<Vec<Vec<Value>>>();
    execute_relation_batch(
        &crate::exec_context::for_driver(driver),
        &op.sql,
        &columns,
        op.key_shape,
        op.hard_limit,
        op.target_table.as_deref(),
        &op.name,
    )
}

fn grouped(op: &RelationOp, rows: Vec<Value>) -> RelationBatch<Value> {
    group_children(rows, |row| {
        op.target_keys
            .iter()
            .map(|key| field(row, key).cloned().unwrap_or(Value::Null))
            .collect()
    })
}

fn distributed(op: &RelationOp, tuple: &[Value], batch: &RelationBatch<Value>) -> Value {
    let rows = matched_children(tuple, batch);
    if op.kind == "hasMany" {
        Value::Arr(rows)
    } else {
        rows.into_iter().next().unwrap_or(Value::Null)
    }
}

pub fn stitch_relation(
    descriptor: &Node,
    mut parents: Vec<Value>,
    driver: &dyn Driver,
) -> Result<Vec<Value>, RuntimeError> {
    let op = op_from_node(descriptor);
    let parent_tuples = tuples(&parents, &op.parent_keys);
    let batch = grouped(&op, fetch(&op, &parent_tuples, driver)?);
    for (parent, tuple) in parents.iter_mut().zip(parent_tuples.iter()) {
        if let Value::Obj(fields) = parent {
            fields.push((op.name.clone(), distributed(&op, tuple, &batch)));
        }
    }
    Ok(parents)
}

pub fn stitch_relation_tree(
    descriptor: &Node,
    parents: Vec<Value>,
    driver: &dyn Driver,
) -> Result<Vec<Value>, RuntimeError> {
    let name = string(descriptor, "name");
    let mut stitched = stitch_relation(descriptor, parents, driver)?;
    let mut children = Vec::new();
    for parent in &stitched {
        if let Some(Value::Arr(rows)) = field(parent, &name) {
            children.extend(rows.iter().cloned());
        }
    }
    if let Some(nested) = descriptor.get("childRelations").and_then(Node::as_array) {
        for child in nested {
            children = stitch_relation_tree(child, children, driver)?;
        }
        let mut hydrated = children.into_iter();
        for parent in &mut stitched {
            if let Some(Value::Arr(rows)) = match parent {
                Value::Obj(fields) => fields
                    .iter_mut()
                    .find(|(field, _)| field == &name)
                    .map(|(_, value)| value),
                _ => None,
            } {
                for row in rows {
                    *row = hydrated.next().expect("relation tree cardinality drift");
                }
            }
        }
    }
    Ok(stitched)
}

fn driver_for<'a>(
    op: &RelationOp,
    primary: &'a (dyn Driver + Sync),
    connections: &'a HashMap<String, &'a (dyn Driver + Sync)>,
) -> Result<&'a (dyn Driver + Sync), SqlFailure> {
    match &op.connection {
        None => Ok(primary),
        Some(name) => connections.get(name).copied().ok_or_else(|| SqlFailure {
            kind: "driver_error".into(),
            policy: "fail".into(),
            sqlite_code: None,
            message: format!(
                "cross-DB relation '{}': connection '{}' is not registered",
                op.name, name
            ),
        }),
    }
}

pub fn read_bundle_pooled(
    bundle: &Node,
    input: &Node,
    driver: &(dyn Driver + Sync),
    with_names: &[String],
    connections: &HashMap<String, &(dyn Driver + Sync)>,
) -> Result<Value, RuntimeError> {
    let mut rows = match execute_bundle(bundle, input, driver)? {
        Value::Arr(rows) => rows,
        _ => {
            return Err(RuntimeError::Sql(SqlFailure {
                kind: "driver_error".into(),
                policy: "fail".into(),
                sqlite_code: None,
                message: "read behavior output is not a row list".into(),
            }))
        }
    };
    for name in with_names {
        let descriptor = bundle
            .get("relations")
            .and_then(|relations| relations.get(name))
            .ok_or_else(|| SqlFailure {
                kind: "driver_error".into(),
                policy: "fail".into(),
                sqlite_code: None,
                message: format!("relation '{name}' is not declared"),
            })?;
        let op = op_from_node(descriptor);
        let relation_driver = driver_for(&op, driver, connections)?;
        rows = stitch_relation_tree(descriptor, rows, relation_driver)?;
    }
    Ok(Value::Arr(rows))
}

/// Single-driver interpreter oracle entry. Kept separate from the pooled cross-DB adapter so native
/// SQLite consumers do not acquire a `Sync` requirement merely to run conformance.
pub fn read_bundle(
    bundle: &Node,
    input: &Node,
    driver: &dyn Driver,
    with_names: &[String],
) -> Result<Value, RuntimeError> {
    let mut rows = match execute_bundle(bundle, input, driver)? {
        Value::Arr(rows) => rows,
        _ => {
            return Err(RuntimeError::Sql(SqlFailure {
                kind: "driver_error".into(),
                policy: "fail".into(),
                sqlite_code: None,
                message: "read behavior output is not a row list".into(),
            }))
        }
    };
    for name in with_names {
        let descriptor = bundle
            .get("relations")
            .and_then(|relations| relations.get(name))
            .ok_or_else(|| SqlFailure {
                kind: "driver_error".into(),
                policy: "fail".into(),
                sqlite_code: None,
                message: format!("relation '{name}' is not declared"),
            })?;
        rows = stitch_relation_tree(descriptor, rows, driver)?;
    }
    Ok(Value::Arr(rows))
}
